docs/examples/self_attentive_sentence_embedding/self_attentive_sentence_embedding.md

```diff
@@ -0,0 +1,535 @@
+
+        dense_out = self.dense(dense_input)
+        output = self.output_layer(F.Dropout(dense_out, self.drop_prob))
+
+        return output, att
+```
+
+## Configure parameters and build models
+
+The resulting `M` is a matrix, and the way to classify this matrix is `flatten`, `mean` or `prune`. Prune is a way of trimming parameters proposed in the original paper and has been implemented here.
+
+```{.python .input}
+vocab_len = len(vocab)
+emsize = 300    # word embedding size
+nhide = 300    # lstm hidden_dim
+nlayers = 2     # lstm layers
+att_unit = 350     # the hidden_units of attenion layer
+att_hops = 2    # the channels of attention
+nfc = 512
+nclass = 5
+
+drop_prob = 0.5
+pool_way = 'flatten'    # # The way to handle M
+prune_p = None
+prune_q = None
+
+ctx = try_gpu()
+
+model = SelfAttentiveBiLSTM(vocab_len, emsize, nhide, nlayers,
+                            att_unit, att_hops, nfc, nclass,
+                            drop_prob, pool_way, prune_p, prune_q)
+
+model.initialize(init=init.Xavier(), ctx=ctx)
+model.hybridize()
+
+# Attach a pre-trained glove word vector to the embedding layer
+model.embedding_layer.weight.set_data(vocab.embedding.idx_to_vec)
+# fixed the layer
+model.embedding_layer.collect_params().setattr('grad_req', 'null')
+```
+
+Using r attention can improve the representation of sentences with different semantics, but if the value of each line in the attention matrix A (r-byn) is very close, that is, there is no difference between several attentions. In the subsequent M = AH, the resulting M will contain a lot of redundant information.
+So in order to solve this problem, we should try to force A to ensure that the value of each line has obvious differences, that is, try to satisfy the diversity of attention. Therefore, a penalty can be used to achieve this goal.
+$$ P = ||(AA^T-I)||_F^2 $$
+
+
+It can be seen from the above formula that if the value of each row of A is more similar, the result of P will be larger, and the value of A is less similar for each row, and P is smaller. This means that when the r-focused diversity of A is larger, the smaller P is. So by including this penalty item with the Loss of the model, you can try to ensure the diversity of A.
+
+```{.python .input}
+def evaluate(data_iter_valid, model, loss, ctx, penal_coeff=0.0, class_weight=None, loss_name='wsce'):
+    valid_loss = 0.
+    total_pred = []
+    total_true = []
+    n_batch = 0
+    for batch_x, batch_y in data_iter_valid:
+        batch_x = batch_x.as_in_context(ctx)
+        batch_y = batch_y.as_in_context(ctx)
+        batch_pred, att = model(batch_x)
+        if loss_name == 'sce':
+            l = loss(batch_pred, batch_y)
+        elif loss_name == 'wsce':
+            l = loss(batch_pred, batch_y, class_weight, class_weight.shape[0])
+
+        # penalty
+        temp = nd.batch_dot(att, nd.transpose(att, axes=(0, 2, 1))
```

Comment:
`temp` is inappropriate since it delivers no information. Why not something like `diversity_penalty`?